### PR TITLE
Add Express middleware.

### PR DESCRIPTION
**Pull request by aomarks**

### CODE REVIEW

File: middleware/.clang-format [hunk 1]
```diff
@@ -0,0 +1,10 @@
+BasedOnStyle: Google
+AlignAfterOpenBracket: AlwaysBreak
+AllowAllParametersOfDeclarationOnNextLine: false
+AllowShortBlocksOnASingleLine: false
+AllowShortCaseLabelsOnASingleLine: false
+AllowShortFunctionsOnASingleLine: None
+AllowShortIfStatementsOnASingleLine: false
+AllowShortLoopsOnASingleLine: false
+BinPackArguments: false
+BinPackParameters: false
```
Comment (samuelli) [inline middleware/.clang-format]: Should this file be committed?

File: middleware/src/middleware.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  'facebookexternalhit',
+  'linkedinbo',
+  'outbrain',
+  'pinterest',
+  'quora link preview',
+  'rogerbo',
+  'showyoubot',
+  'slackbot',
+  'twitterbot',
+  'vkShare',
```
Comment (samuelli) [inline middleware/src/middleware.js]: Do you think we should pull this out and publish it separately? Either as a JSON file, or node module.

File: middleware/src/middleware.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ */
+module.exports.makeMiddleware = function(options) {
+  if (!options || !options.proxyURL) {
+    throw new Error('Must set options.proxyURL.');
+  }
+  const proxyURL = options.proxyURL;
+  const userAgentMatcher =
+      options.userAgentMatcher || new RegExp(botUserAgents.join('|'), 'i');
+  const injectShadyDOM = !!options.injectShadyDOM;
+  const timeout = options.timeout || 10000;  // Milliseconds.
```
Comment (samuelli) [inline middleware/src/middleware.js]: Use something > 10s, the service has a hard limit of 10s to render, so + some extra latency to wait for that.

File: middleware/src/middleware.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+  return function rendertronMiddleware(req, res, next) {
+    if (!userAgentMatcher.test(req.headers['user-agent'])) {
+      next();
+      return;
+    }
```
Comment (samuelli) [inline middleware/src/middleware.js]: Are we going to look at the resource type in a followup PR?

File: .travis.yml [hunk 5]
```diff
@@ -12,9 +12,11 @@ install:
     fi
   - ${HOME}/gcloud/google-cloud-sdk/bin/gcloud components install beta cloud-datastore-emulator
   - source $HOME/gcloud/google-cloud-sdk/path.bash.inc
+  - cd middleware && yarn
```
Comment (samuelli) [inline .travis.yml]: Looks like this cd stuck around.
Change to
```
(cd middleware && yarn)
```

File: middleware/src/middleware.js (hunk 2 above)
Comment (aomarks) [inline middleware/src/middleware.js]: Maybe later?

File: middleware/src/middleware.js (hunk 3 above)
Comment (aomarks) [inline middleware/src/middleware.js]: 11s

File: middleware/.clang-format (hunk 1 above)
Comment (aomarks) [inline middleware/.clang-format]: I think so. It means anybody who edits this file and has clang-format will automatically abide by our style. This is what we do in all the tools team projects.

File: middleware/src/middleware.js (hunk 4 above)
Comment (aomarks) [inline middleware/src/middleware.js]: Added to this PR.

File: .travis.yml (hunk 5 above)
Comment (aomarks) [inline .travis.yml]: Done.

**Comment by aomarks** [comment]
PTAL, just added a bunch of tests and fixed a couple bugs I found.

File: middleware/test/middleware-test.js [hunk 6]
```diff
@@ -1,15 +1,158 @@
...
+      .use((req, res) => res.end('fallthrough'));
+}
+
+/**
+ * Make an Express app that takes the place of a Rendertron server instance and
+ * always responds with "proxy <decoded url>".
+ */
+function makeProxy() {
+  return express().use((req, res) => {
+    res.end('proxy ' + decodeURIComponent(req.url.substring(1)));
```
Comment (samuelli) [inline middleware/test/middleware-test.js]: Why is substring needed?

**Pull request merged**